### ISSUE TITLE
On-demand server-timing

### ISSUE THREAD

**Issue by steve-chavez** [PostgREST/postgrest#3410]
## Problem

Setting [server-timing-enabled](https://postgrest.org/en/v12/references/configuration.html#server-timing-enabled) to true has  a ~6.5% RPS decrease. Using GETSimple on https://github.com/PostgREST/postgrest-benchmark:

no timings = 2176.4712194 rps
with timings = 2033.3584706 rps

I don't think there's a way around it since calculating the timings requires using clock time which needs a system call (`clock_gettime`).

## Solution

Allow getting server timing headers through a new preference header.

```http
Prefer: metrics=timings
```

`metrics=timings` would be the only value for now, but we could extend it later on.

This can later be enabled/disabled with https://github.com/PostgREST/postgrest/issues/2987.

While at it, deprecate [server-timing-enabled](https://postgrest.org/en/v12/references/configuration.html#server-timing-enabled) since it has a perf impact.

**Comment by develop7** [comment]
> Setting [server-timing-enabled](https://postgrest.org/en/v12/references/configuration.html#server-timing-enabled) to true has a ~6.5% RPS decrease.

Damn, I wonder how much slower will OTel be when enabled then.

**Comment by steve-chavez** [comment]
Yeah, this makes sense since `server-timing` is a trace and by definition:

>  Running traces takes too much time and consumes too many resources to trace every request an application receives.
> https://www.techtarget.com/searchitoperations/tip/The-3-pillars-of-observability-Logs-metrics-and-traces

> Damn, I wonder how much slower will OTel be when enabled then.

Not yet familiar with OTel (and the Push model) but doesn't it do some buffering before sending the events to a collector?

I've also read that they do [sampling](https://opentelemetry.io/docs/concepts/sampling/), so not all requests are traced. This would reduce resource consumption.

---

It would be interesting to have an internal sampling mechanism, so we could have some server-timing traces exposed as metrics. (just an idea for now, not sure if worth it)

**Comment by taimoorzaeem** [comment]
@steve-chavez As I am working on this,
```haskell
middleware :: AppState -> Wai.Middleware
middleware appState app req respond = do
  conf <- getConfig appState
  time <- getTime appState

  let token  = fromMaybe "" $ Wai.extractBearerAuth =<< lookup HTTP.hAuthorization (Wai.requestHeaders req)
      parseJwt = runExceptT $ parseToken conf (LBS.fromStrict token) time >>= parseClaims conf

-- If ServerTimingEnabled  -> calculate JWT validation time
-- If JwtCacheMaxLifetime -> cache JWT validation result
  req' <- case (configServerTimingEnabled conf, configJwtCacheMaxLifetime conf) of
    (True, 0)            -> do
```
How do I get `ApiRequest` object into `Auth.middleware`? I need the `ApiRequest` to replace the `configServerTimingEnabled` with the new preference but I think it is not possible here. WDYT?

**Comment by taimoorzaeem** [comment]
@steve-chavez Do we really need the `Preference-Applied: ...` for this? I guess not. 

How about we solve this with a custom header instead of a preference?  It would be easier to parse it in `Auth.hs` module because it would just a be simple header lookup. Maybe something like `X-Metrics: timings`? Later, we can extend it to other values too?

**Comment by steve-chavez** [comment]
Hm, we always frown upon using non-standard headers as it adds complexity (for us and the user) plus the `Auth.hs` module doesn't seem like the right place to do this. I think this should wait until the proposed refactor in https://github.com/PostgREST/postgrest/pull/4064#issuecomment-2859965093 is done.
